File: tests/BeaverTests/Examples/Arguments.swift
import Testing
import Foundation
import Darwin
import Utils
@testable import Beaver

@Test func exampleArguments() async throws {
  let baseURL = URL(filePath: "Examples/Arguments", relativeTo: URL.currentDirectory())

  var (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["helloWorld"], baseDir: baseURL)
  #expect(stdout == "Hello world\n")

  (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["shellCommand"], baseDir: baseURL)
  #expect(stdout == "Hello world!\nHello world!\n")

  (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["printArg", "--argument-name", "warn", "--warn"], baseDir: baseURL)
  #expect(stdout == "true\n")

  (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["printArg", "--argument-name", "warn", "-w"], baseDir: baseURL)
  #expect(stdout == "true\n")

  (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["printArg", "--argument-name", "warn", "--no-warn"], baseDir: baseURL)
  #expect(stdout == "false\n")

  (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["printArg", "--argument-name", "warn"], baseDir: baseURL)
  #expect(stdout == "\n")

  (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["printArg", "--argument-name", "sdl-version"], baseDir: baseURL)
  #expect(stdout == "2\n")

  (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["printArg", "--argument-name", "sdl-version", "--sdl-version", "3"], baseDir: baseURL)
  #expect(stdout == "3\n")

  (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["printArg", "--argument-name", "sdl-version", "-s", "1"], baseDir: baseURL)
  #expect(stdout == "1\n")

  (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["printArg", "--argument-name", "sdl-version"], baseDir: baseURL)
  #expect(stdout == "2\n")

  (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["printArg", "--argument-name", "from-source"], baseDir: baseURL)
  #expect(stdout == "false\n")

  (stdout, _) = try Tools.execWithOutput(beaverExeURL, ["printArg", "--argument-name", "from-source", "--from-source"], baseDir: baseURL)
  #expect(stdout == "true\n")
}

File: beaver-cli/tests/examples/arguments.rs
use std::process::Command;

#[test]
fn opt() {
    let dir = super::example_dir("arguments");
    let output = Command::new(crate::beaver())
        .args(["printArg", "--", "--argument-name", "sdl-version", "--sdl-version", "3"])
        .current_dir(dir)
        .output()
        .unwrap();

    assert_eq!(output.status.code(), Some(0));
    assert_eq!(String::from_utf8(output.stdout).unwrap(), "3\n");
}

#[test]
fn opt_short() {
    let dir = super::example_dir("arguments");
    let output = Command::new(crate::beaver())
        .args(["printArg", "--", "--argument-name", "sdl-version", "-s", "3"])
        .current_dir(dir)
        .output()
        .unwrap();

    assert_eq!(output.status.code(), Some(0));
    assert_eq!(String::from_utf8(output.stdout).unwrap(), "3\n");
}

#[test]
fn opt_default() {
    let dir = super::example_dir("arguments");
    let output = Command::new(crate::beaver())
        .args(["printArg", "--", "--argument-name", "sdl-version"])
        .current_dir(dir)
        .output()
        .unwrap();

    assert_eq!(output.status.code(), Some(0));
    assert_eq!(String::from_utf8(output.stdout).unwrap(), "2\n");
}

#[test]
fn flag() {
    let dir = super::example_dir("arguments");
    let output = Command::new(crate::beaver())
        .args(["printArg", "--", "--argument-name", "from-source", "--from-source"])
        .current_dir(dir)
        .output()
        .unwrap();

    assert_eq!(output.status.code(), Some(0));
    assert_eq!(String::from_utf8(output.stdout).unwrap(), "true\n");
}

#[test]
fn flag_default() {
    let dir = super::example_dir("arguments");
    let output = Command::new(crate::beaver())
        .args(["printArg", "--", "--argument-name", "from-source"])
        .current_dir(dir)
        .output()
        .unwrap();

    assert_eq!(output.status.code(), Some(0));
    assert_eq!(String::from_utf8(output.stdout).unwrap(), "false\n");
}

#[test]
fn flag_nil_default() {
    let dir = super::example_dir("arguments");
    let output = Command::new(crate::beaver())
        .args(["printArg", "--", "--argument-name", "warn"])
        .current_dir(dir)
        .output()
        .unwrap();

    assert_eq!(output.status.code(), Some(0));
    assert_eq!(String::from_utf8(output.stdout).unwrap(), "\n");
}

#[test]
fn flag_nil_present() {
    let dir = super::example_dir("arguments");
    let output = Command::new(crate::beaver())
        .args(["printArg", "--", "--argument-name", "warn", "--warn"])
        .current_dir(dir)
        .output()
        .unwrap();

    assert_eq!(output.status.code(), Some(0));
    assert_eq!(String::from_utf8(output.stdout).unwrap(), "true\n");
}

#[test]
fn flag_nil_negated() {
    let dir = super::example_dir("arguments");
    let output = Command::new(crate::beaver())
        .args(["printArg", "--", "--argument-name", "warn", "--no-warn"])
        .current_dir(dir)
        .output()
        .unwrap();

    assert_eq!(output.status.code(), Some(0));
    assert_eq!(String::from_utf8(output.stdout).unwrap(), "false\n");
}

#[test]
fn cmd() {
    let dir = super::example_dir("arguments");
    let output = Command::new(crate::beaver())
        .args(["helloWorld"])
        .current_dir(dir)
        .output()
        .unwrap();

    assert_eq!(output.status.code(), Some(0));
    assert_eq!(String::from_utf8(output.stdout).unwrap(), "Hello world\n");
}

#[test]
fn cmd_shell() {
    let dir = super::example_dir("arguments");
    let output = Command::new(crate::beaver())
        .args(["shellCommand"])
        .current_dir(dir)
        .output()
        .unwrap();

    assert_eq!(output.status.code(), Some(0));
    assert_eq!(String::from_utf8(output.stdout).unwrap(), "Hello world!\nHello world!\n");
    assert!(output.stderr.len() > 1);
}
